### PR TITLE
Expose `ArrayElement` type

### PR DESCRIPTION
**Pull request by tommy-mitchell**
Related: #676. (maybe closes?)

Exposes internal `ArrayElement` type:

> Extracts the element type of an array or union of arrays.

```ts
import type {ArrayElement} from 'type-fest';

declare const getMostCommonElement: <T>(array: T[]) => ArrayElement<typeof array>;

getMostCommonElement([1, 2, 3]);
//=> 1 | 2 | 3

getMostCommonElement(['foo', 'bar', 'baz'] as const);
//=> 'foo' | 'bar' | 'baz'
```

### CODE REVIEW

File: source/array-element.d.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+```
+
+@see {@link ArrayValues} for when you know that the input is an array.
+
+@category Array
+*/
+export type ArrayElement<ArrayType extends readonly unknown[]> =
+	ArrayType extends ReadonlyArray<infer ElementType>
+		? ElementType
+		: never;
```
Comment (tommy-mitchell) [inline source/array-element.d.ts:28]: I'm not sure this works @strongpauly

> Should this be more?:
> 
> ```ts
> type ArrayElement<ArrayType extends readonly unknown[]> = 
>   ArrayType extends readonly (infer ElementType)[] ? ElementType : never;
> ```
> 
> Copy paste from [stackoverflow.com/questions/41253310/typescript-retrieve-element-type-information-from-array-type](https://stackoverflow.com/questions/41253310/typescript-retrieve-element-type-information-from-array-type)

// https://github.com/sindresorhus/type-fest/pull/906#discussion_r1673712673

**Comment by sindresorhus** [comment]
Isn't this type already covered by https://github.com/sindresorhus/type-fest/blob/main/source/iterable-element.d.ts ?

File: source/array-element.d.ts (hunk 1 above)
Comment (strongpauly) [reply source/array-element.d.ts:28]: Well, as long as it returns a union of the types of the array elements, its fit for purpose, IMO.

**Comment by tommy-mitchell** [comment]
> Isn't this type already covered by [`main`/source/iterable-element.d.ts](https://github.com/sindresorhus/type-fest/blob/main/source/iterable-element.d.ts?rgh-link-date=2024-07-14T10%3A56%3A22Z) ?

Hm, I'd assume so, but the tests for `Exact` fail if swapped to use `IterableElement`:

image

**Comment by fregante** [comment]
Is it really worth to import and use a custom type instead of just using a literal type offered by TypeScript?

``` diff
- array[number]
+ ArrayItem<array>
```

**Comment by fregante** [comment]
In your example it does not look like an advantage, it's hard to follow and does not match the order of the JS code.


```ts
              0             1            2        3 
              complexObject .property_1  [0].     .subprop_1

              0             1            2        3
type MyType = MyComplexType['property_1'][number]['subprop_1'] | null
// vs

              2           0             1              3
type MyType = ArrayElement<MyComplexType['property_1']>['subprop_1'] | null

```

File: source/array-element.d.ts (hunk 1 above)
Comment (giao66) [reply source/array-element.d.ts:28]: good
